### PR TITLE
Table Definitions and Route Definitions

### PR DESCRIPTION
**Pull request by hoguer**
Table Definitions and Initial Route Definitions

### CODE REVIEW

File: db/init_db.js [hunk 1]
```diff
@@ -1,25 +1,477 @@
 // code to build and initialize DB goes here
 const {
   client
```
Comment (hoguer) [inline db/init_db.js:3]: it looks like we need to update this to import from client instead of index

File: db/init_db.js [hunk 2]
```diff
@@ -1,25 +1,477 @@
...
+        DROP TABLE IF EXISTS users, products, orders, order_products;
+      `);
+  
+      await client.query(`
+      CREATE TABLE users(
+        id SERIAL PRIMARY KEY, 
+        firstname VARCHAR(255) NOT NULL, 
+        lastname VARCHAR(255) NOT NULL, 
+        email VARCHAR(255) UNIQUE NOT NULL, 
+        "imgURL" DEFAULT VALUE,
```
Comment (hoguer) [inline db/init_db.js]: let's get the actual default VALUE in here to start, something like:
`"imgURL" DEFAULT default_img.jpg,`

File: db/init_db.js [hunk 3]
```diff
@@ -1,25 +1,477 @@
...
+        address NOT NULL
+      );
+    `);
+    await client.query(`
+    CREATE TABLE products(
+      id SERIAL PRIMARY KEY, 
+      name VARCHAR(255) UNIQUE NOT NULL, 
+      description STRING, 
+      price INTEGER NOT NULL, 
+      "imgURL" DEFAULT VALUE,
```
Comment (hoguer) [inline db/init_db.js]: same thing here; let's get an actual value to replace `VALUE`

File: db/init_db.js [hunk 4]
```diff
@@ -1,25 +1,477 @@
...
+    `);
+    await client.query(`
+    CREATE TABLE products(
+      id SERIAL PRIMARY KEY, 
+      name VARCHAR(255) UNIQUE NOT NULL, 
+      description STRING, 
+      price INTEGER NOT NULL, 
+      "imgURL" DEFAULT VALUE,  
+      "isStock" BOOLEAN DEFAULT true,
+      category NOT NULL
```
Comment (hoguer) [inline db/init_db.js]: specify category type

File: db/init_db.js [hunk 5]
```diff
@@ -1,25 +1,477 @@
...
+  `);
 
-    // drop tables in correct order
-
-    // build tables in correct order
+  await client.query(`
+  CREATE TABLE orders(
+    id SERIAL PRIMARY KEY, 
+    status DEFAULT "created", 
+    “userID” INTEGER REFERENCES users(id),
```
Comment (hoguer) [inline db/init_db.js]: it might complain about the slanty quotes.

File: db/init_db.js [hunk 6]
```diff
@@ -1,25 +1,477 @@
...
 
-    // drop tables in correct order
-
-    // build tables in correct order
+  await client.query(`
+  CREATE TABLE orders(
+    id SERIAL PRIMARY KEY, 
+    status DEFAULT "created", 
+    “userID” INTEGER REFERENCES users(id), 
+    "datePlaced" date
```
Comment (hoguer) [inline db/init_db.js]: you can default date types to the current time

File: db/init_db.js [hunk 7]
```diff
@@ -1,25 +1,477 @@
...
+
 
 async function populateInitialData() {
   try {
     // create useful starting data
+    console.log("populating our wine and cheese tables")
+
+    const wineAndCheeseData = [
+      {
+        productId: 1,
```
Comment (hoguer) [inline db/init_db.js]: don't need the id since it's auto-incrementing

File: db/products.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    } catch (error) {
+        throw error;
+    };
+};
+​
+async function getAllProducts() {
+    try {
+        const {rows} = await client.query(`
+            SELECT * FROM products
+        `, []);
```
Comment (hoguer) [inline db/products.js]: you can exclude the second parameter when it's an empty array.

File: db/products.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    } catch (error) {
+        throw error;
+    };
+};
+
+async function getProductByName(name) {
+    try {
+        const {rows: [product]} = await client.query(`
+            SELECT * FROM products 
+            WHERE id=$1
```
Comment (hoguer) [inline db/products.js]: this WHERE should look for name, not id

File: db/users.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
+const client = require("../db/client");
```
Comment (hoguer) [inline db/users.js]: can just be `const client = require("./client");`

File: db/products.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
+const client = require("../db/client");
```
Comment (hoguer) [inline db/products.js]: can just be `const client = require("./client");`

File: db/users.js [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+const bcrypt = require("bcrypt");
+const SALT_COUNT = 10;
+
+async function createUser ({ username, password }) {
```
Comment (hoguer) [inline db/users.js]: we need to pass in the other user data that is used on line 9

File: db/users.js [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    const hashedPassword = await bcrypt.hash(password, SALT_COUNT);
+    try {
+        const { rows: [user] } = await client.query(`
+        INSERT INTO users(firstname, lastname, email, "imgURL", username, password, "isAdmin", address) 
+        VALUES ($1, $2)
+        ON CONFLICT (username) DO NOTHING
+        RETURNING id, username
+        `, [username, hashedPassword]);
+    return user;
```
Comment (hoguer) [inline db/users.js]: indent once

File: db/users.js [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    delete user.password;
+    return user;
+} catch (error) {
+    throw error;
+  }
+}
+
+async function getUserByUsername(userName){
+    try{
+        const {rows: [user], }= await client.query(`
```
Comment (hoguer) [inline db/users.js]: can get rid of comma after `[user]`

File: db/users.js [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+      const hashedPassword = user.password;
+      const matchedPass = await bcrypt.compare(password, hashedPassword);
+  
+      if (matchedPass) {
+        console.log('matched')
+        delete user.password;
+        return user;
+      } else {
+        console.log('didnt match')
+        return "";
```
Comment (hoguer) [inline db/users.js]: return null for consistency

**Comment by richardedwardbrown3** [comment]
Looked over. I trust this one.

**Pull request merged**